mlcg/nn/prior.py

```diff
@@ -1065,3 +1065,233 @@ def from_user(*args):
+
+        self.n_degs = n_degs
+        self.k_names = ["k_" + str(ii) for ii in range(2, self.n_degs + 1)]
+        self.x0_names = ["x0_" + str(ii) for ii in range(2, self.n_degs + 1)]
+
+        k = torch.zeros(self.n_degs - 1, *sizes)
+        x0 = torch.zeros(self.n_degs - 1, *sizes)
+        v_0 = torch.zeros(*sizes)
+
+        for key in statistics.keys():
+            for ii in range(self.n_degs - 1):
+                k_name = self.k_names[ii]
+                x0_name = self.x0_names[ii]
+                k[ii][key] = statistics[key]["ks"][k_name]
+                x0[ii][key] = statistics[key]["x0s"][x0_name]
+            v_0[key] = statistics[key]["v_0"]
+        self.register_buffer("ks", k)
+        self.register_buffer("v_0", v_0)
+        self.register_buffer("x0s", x0)
+
+    @staticmethod
+    def compute_features(pos, mapping, target):
+        compute_map_type = Quartic._neighbor_list_map[target]
+        return Quartic._compute_map[compute_map_type](pos, mapping)
+
+    def data2features(self, data):
+        mapping = data.neighbor_list[self.name]["index_mapping"]
+        return Quartic.compute_features(data.pos, mapping, self.name)
+
+    def data2parameters(self, data):
+        mapping = data.neighbor_list[self.name]["index_mapping"]
+        interaction_types = [
+            data.atom_types[mapping[ii]] for ii in range(self.order)
+        ]
+        # the parameters have shape n_features x n_degs-1 since
+        # linear term is missing
+        ks = torch.vstack(
+            [self.ks[ii][interaction_types] for ii in range(self.n_degs - 1)]
+        ).t()
+
+        x0s = torch.vstack(
+            [self.x0s[ii][interaction_types] for ii in range(self.n_degs - 1)]
+        ).t()
+        v_0s = self.v_0[interaction_types].t()
+        return {"ks": ks, "x0s": x0s, "v_0s": v_0s}
+
+    def forward(self, data):
+        mapping_batch = data.neighbor_list[self.name]["mapping_batch"]
+        features = self.data2features(data).flatten()
+        params = self.data2parameters(data)
+        V0s = params["v_0s"].t()
+        ks = params["ks"].t()
+        x0s = params["x0s"].t()
+        y = Quartic.compute(
+            features,
+            ks,
+            V0s,
+            x0s,
+        )
+        y = scatter(y, mapping_batch, dim=0, reduce="sum")
+        data.out[self.name] = {"energy": y}
+        return data
+
+    @staticmethod
+    def compute(
+        x: torch.Tensor, ks: torch.Tensor, V0: torch.Tensor, x0s: torch.Tensor
+    ):
+        """Quartic potential interaction with missing linear term.
+
+        .. math:
+
+            V(r) = V0 + \sum_{n=2}^{4} k_n (x-x_n)^n
+
+        """
+        V = 0
+        for i in range(3):
+            V += ks[i] * (x - x0s[i]) ** (i + 2)
+
+        V += V0
+        return V
+
+    @staticmethod
+    def _quartic_model(x, a, b, c, d, xa, xb, xc):
+        return a * (x - xa) ** 2 + b * (x - xb) ** 3 + c * (x - xc) ** 4 + d
+
```

Comment:
A deg 4 polynomial p(x) is uniquely caracterized by 5 numbers: a_0, .. a_4 such that p(x) = sum_{k=0}^4 a_k * x^k. Your degree 4 polynomial uses 7 parameters in order to keep this shape of  p(x) = sum_{k=2}^4 b_k *(x_0^{(k)} x)^k. Isn't the fit then ill posed?  

At some point I tried to use an expression similar to what you have now and found it problematic because of this redundancy of parameters.